File: foundational_design_patterns/3_parallelization/src/parallelization.py
"""
Parallel Chain Processing Example

This module demonstrates parallel execution of multiple LangChain tasks
(summarization, question generation, and key term extraction) followed by
synthesis of results into a comprehensive output.
"""

import asyncio
import os
from typing import Dict, Any, Optional
from dotenv import load_dotenv

from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import Runnable, RunnableParallel, RunnablePassthrough

# ============================================================================
# CONFIGURATION
# ============================================================================

# Load environment variables
load_dotenv(os.path.join(os.path.dirname(__file__), "../../..", ".env"))

# LLM Configuration
LLM_MODEL = "gpt-4o-mini"
LLM_TEMPERATURE = 0.7

# Initialize LLM
llm = ChatOpenAI(model=LLM_MODEL, temperature=LLM_TEMPERATURE)


# ============================================================================
# CHAIN DEFINITIONS
# ============================================================================

def create_summarize_chain() -> Runnable:
    """Create a chain that generates a concise summary of a topic."""
    prompt = ChatPromptTemplate.from_messages([
        ("system", "Summarize the following topic concisely in 2-3 sentences:"),
        ("user", "{topic}")
    ])
    return prompt | llm | StrOutputParser()


def create_questions_chain() -> Runnable:
    """Create a chain that generates interesting questions about a topic."""
    prompt = ChatPromptTemplate.from_messages([
        (
            "system",
            "Generate three thought-provoking questions about the following topic. "
            "Format each question on a new line with a number."
        ),
        ("user", "{topic}")
    ])
    return prompt | llm | StrOutputParser()


def create_terms_chain() -> Runnable:
    """Create a chain that extracts key terms from a topic."""
    prompt = ChatPromptTemplate.from_messages([
        (
            "system",
            "Identify 5-10 key terms or concepts related to the following topic. "
            "Return them as a comma-separated list."
        ),
        ("user", "{topic}")
    ])
    return prompt | llm | StrOutputParser()


def create_synthesis_chain() -> Runnable:
    """Create a chain that synthesizes parallel results into a comprehensive output."""
    prompt = ChatPromptTemplate.from_messages([
        (
            "system",
            """You are synthesizing information about a topic. Based on the following:

Summary: {summary}

Related Questions:
{questions}

Key Terms: {key_terms}

Create a comprehensive, well-structured response that:
1. Integrates the summary with the key terms naturally
2. Addresses or contextualizes the questions raised
3. Provides additional insights or connections
4. Maintains clarity and coherence"""
        ),
        ("user", "Original topic: {topic}")
    ])
    return prompt | llm | StrOutputParser()


# ============================================================================
# PARALLEL CHAIN CONSTRUCTION
# ============================================================================

def build_parallel_chain() -> Runnable:
    """
    Build a complete parallel processing chain.
    
    Returns:
        A Runnable that executes summarization, question generation, and
        term extraction in parallel, then synthesizes the results.
    """
    # Create independent chains
    summarize_chain = create_summarize_chain()
    questions_chain = create_questions_chain()
    terms_chain = create_terms_chain()
    
    # Define parallel execution block
    map_chain = RunnableParallel({
        "summary": summarize_chain,
        "questions": questions_chain,
        "key_terms": terms_chain,
        "topic": RunnablePassthrough(),
    })
    
    # Create synthesis chain
    synthesis_chain = create_synthesis_chain()
    
    # Combine into full pipeline
    full_chain = map_chain | synthesis_chain | StrOutputParser()
    
    return full_chain


# ============================================================================
# EXECUTION FUNCTIONS
# ============================================================================

async def process_topic_async(topic: str, chain: Optional[Runnable] = None) -> Dict[str, Any]:
    """
    Asynchronously process a topic through the parallel chain.
    
    Args:
        topic: The input topic to be processed
        chain: Optional pre-built chain (creates new one if None)
        
    Returns:
        Dictionary containing the final response and metadata
        
    Raises:
        Exception: If chain execution fails
    """
    if chain is None:
        chain = build_parallel_chain()
    
    print(f"\n{'=' * 70}")
    print(f"Processing Topic: '{topic}'")
    print(f"{'=' * 70}\n")
    
    try:
        # Execute the chain
        response = await chain.ainvoke(topic)
        
        result = {
            "topic": topic,
            "response": response,
            "status": "success"
        }
        
        print("\n--- Final Synthesized Response ---")
        print(response)
        print(f"\n{'=' * 70}\n")
        
        return result
        
    except Exception as e:
        error_msg = f"Chain execution failed: {str(e)}"
        print(f"\n❌ ERROR: {error_msg}\n")
        
        return {
            "topic": topic,
            "response": None,
            "status": "error",
            "error": error_msg
        }


def process_topic_sync(topic: str) -> Dict[str, Any]:
    """
    Synchronously process a topic (wrapper for async function).
    
    Args:
        topic: The input topic to be processed
        
    Returns:
        Dictionary containing the final response and metadata
    """
    return asyncio.run(process_topic_async(topic))


async def process_multiple_topics(topics: list[str]) -> list[Dict[str, Any]]:
    """
    Process multiple topics in parallel.
    
    Args:
        topics: List of topics to process
        
    Returns:
        List of result dictionaries for each topic
    """
    chain = build_parallel_chain()
    tasks = [process_topic_async(topic, chain) for topic in topics]
    return await asyncio.gather(*tasks)


# ============================================================================
# MAIN EXECUTION
# ============================================================================

def main():
    """Main execution function with example usage."""
    
    # Single topic example
    print("\n🚀 Running Single Topic Example")
    single_topic = "The history of space exploration"
    result = process_topic_sync(single_topic)
    
    # Multiple topics example (commented out by default)
    # print("\n🚀 Running Multiple Topics Example")
    # multiple_topics = [
    #     "The history of space exploration",
    #     "Quantum computing fundamentals",
    #     "Climate change mitigation strategies"
    # ]
    # results = asyncio.run(process_multiple_topics(multiple_topics))
    # 
    # # Print summary
    # successful = sum(1 for r in results if r["status"] == "success")
    # print(f"\n✅ Successfully processed {successful}/{len(results)} topics")


if __name__ == "__main__":
    main()

File: foundational_design_patterns/3_parallelization/typescript/src/parallelization.ts
/**
 * Parallelization Pattern — TypeScript port of src/parallelization.py.
 */

import { ChatOpenAI } from "@langchain/openai";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { StringOutputParser } from "@langchain/core/output_parsers";
import { RunnableMap, RunnablePassthrough } from "@langchain/core/runnables";

const LLM_MODEL = process.env.OPENAI_MODEL ?? "gpt-4o-mini";
const LLM_TEMPERATURE = 0.7;

const llm = new ChatOpenAI({
  model: LLM_MODEL,
  temperature: LLM_TEMPERATURE,
});

export function createSummarizeChain() {
  const prompt = ChatPromptTemplate.fromMessages([
    ["system", "Summarize the following topic concisely in 2-3 sentences:"],
    ["user", "{topic}"],
  ]);
  return prompt.pipe(llm).pipe(new StringOutputParser());
}

export function createQuestionsChain() {
  const prompt = ChatPromptTemplate.fromMessages([
    [
      "system",
      "Generate three thought-provoking questions about the following topic. Format each question on a new line with a number.",
    ],
    ["user", "{topic}"],
  ]);
  return prompt.pipe(llm).pipe(new StringOutputParser());
}

export function createTermsChain() {
  const prompt = ChatPromptTemplate.fromMessages([
    [
      "system",
      "Identify 5-10 key terms or concepts related to the following topic. Return them as a comma-separated list.",
    ],
    ["user", "{topic}"],
  ]);
  return prompt.pipe(llm).pipe(new StringOutputParser());
}

export function createSynthesisChain() {
  const prompt = ChatPromptTemplate.fromMessages([
    [
      "system",
      `You are synthesizing information about a topic. Based on the following:

Summary: {summary}

Related Questions:
{questions}

Key Terms: {key_terms}

Create a comprehensive, well-structured response that:
1. Integrates the summary with the key terms naturally
2. Addresses or contextualizes the questions raised
3. Provides additional insights or connections
4. Maintains clarity and coherence`,
    ],
    ["user", "Original topic: {topic}"],
  ]);
  return prompt.pipe(llm).pipe(new StringOutputParser());
}

export function buildParallelChain() {
  const summarizeChain = createSummarizeChain();
  const questionsChain = createQuestionsChain();
  const termsChain = createTermsChain();

  const mapChain = RunnableMap.from<string>({
    summary: async (topic: string) =>
      await summarizeChain.invoke({ topic }),
    questions: async (topic: string) =>
      await questionsChain.invoke({ topic }),
    key_terms: async (topic: string) =>
      await termsChain.invoke({ topic }),
    topic: new RunnablePassthrough(),
  });

  const synthesisChain = createSynthesisChain();
  return mapChain.pipe(synthesisChain).pipe(new StringOutputParser());
}

export type ProcessResult = {
  topic: string;
  response: string | null;
  status: "success" | "error";
  error?: string;
};

export async function processTopicAsync(
  topic: string,
  chain = buildParallelChain(),
): Promise<ProcessResult> {
  console.log(`\n${"=".repeat(70)}`);
  console.log(`Processing Topic: '${topic}'`);
  console.log(`${"=".repeat(70)}\n`);

  try {
    const response = await chain.invoke(topic);
    console.log("\n--- Final Synthesized Response ---");
    console.log(response);
    console.log(`\n${"=".repeat(70)}\n`);
    return { topic, response, status: "success" };
  } catch (e: unknown) {
    const errorMsg = `Chain execution failed: ${
      e instanceof Error ? e.message : String(e)
    }`;
    console.log(`\n❌ ERROR: ${errorMsg}\n`);
    return {
      topic,
      response: null,
      status: "error",
      error: errorMsg,
    };
  }
}

export async function processMultipleTopics(
  topics: string[],
): Promise<ProcessResult[]> {
  const chain = buildParallelChain();
  const tasks = topics.map((topic) => processTopicAsync(topic, chain));
  return await Promise.all(tasks);
}

export function processTopicSync(topic: string): Promise<ProcessResult> {
  return processTopicAsync(topic);
}

async function main(): Promise<void> {
  console.log("\n🚀 Running Single Topic Example");
  const singleTopic = "The history of space exploration";
  await processTopicSync(singleTopic);

  // console.log("\n🚀 Running Multiple Topics Example");
  // const multipleTopics = [
  //   "The history of space exploration",
  //   "Quantum computing fundamentals",
  //   "Climate change mitigation strategies",
  // ];
  // const results = await processMultipleTopics(multipleTopics);
  // const successful = results.filter((r) => r.status === "success").length;
  // console.log(`\n✅ Successfully processed ${successful}/${results.length} topics`);
}

if (import.meta.main) {
  await main();
}
